net: refactor the connection process. moving towards async connections.

theuni opened the pull request:
This untangles our current connection logic and moves toward async connections/resolves. Before merging, I'd like to construct a matrix of different connection configs, and make sure that each one has test coverage. I'm PRing now to get a feel for whether this is reviewable as-is or not.

In short, switch from "here's what we know, connect please!" to "walk through the individual connection steps in-order".

This moves the entire process into one function, but the next step will be to break out the individual steps (input checking, resolving, connection, node creation).

See the individual commit messages for more detail.

theuni committed:
net: pass a reference into ConnectNode and copy it there

No functional change, just preparation for the following commits.

theuni committed:
net: combine OpenNetworkConnection and ConnectNode

ConnectNode is only called by OpenNetworkConnection, and only in one place.
Combine them for preparation of splitting them into functions that are
friendlier for callbacks.

theuni committed:
net: split socket creation out of connection

theuni committed:
net: expose a few functions/structs from netbase

The next commits will move the connection logic out of netbase into net.

theuni committed:
net: refactor the outgoing connection process

This is the next step towards asynchronous connections.

The current code confuses connection steps and tends to obfuscate what's going
on. Here, I've tried to create individual stages. The logic has all moved into
a single function for now:

- Global checks: Should any connection be attempted?
- Input checks: is there something valid to connect to?
- Resolving: If the address string may be valid, attempt to resolve it (if
    allowed by the user's dns settings and proxy config)
- Proxy config
- Socket creation
- Connect to the destination
- If proxied, attempt the remote connection
- Create the node with the resulting connection

With this done, resolves and connections can trigger callbacks so that a
connection attempt requires no return value. At that point, they can be made
asynchronous via libevent.

theuni committed:
net: remove now-unused functions

gmaxwell commented:
Vague Concept ACK.

laanwj commented:
utACK, going to test this

TheBlueMatt commented on src/netbase.cpp:
```diff
@@ -445,6 +439,19 @@ bool static ConnectSocketDirectly(const CService &addrConnect, SOCKET& hSocketRe
     if (!SetSocketNonBlocking(hSocket, true))
         return error("ConnectSocketDirectly: Setting socket to non-blocking failed, error %s\n", NetworkErrorString(WSAGetLastError()));
```
nit: needs a string change here for \_\_func__.

TheBlueMatt replied:
Oh, also think we need a CloseSocket here.

theuni replied:
Will fix __func__.

SetSocketNonBlocking should take care of closing the socket, but an explicit one here wouldn't hurt either.

CloseSocket is really annoying. I'll RAII it directly after this PR.

TheBlueMatt replied:
Good, CloseSocket were half the issues here :(

TheBlueMatt commented on src/net.cpp:
```diff
@@ -340,71 +340,161 @@ bool CConnman::CheckIncomingNonce(uint64_t nonce)
+    addrman.Attempt(addrDest, fCountFailure);
+
+    if (!connected) {
+        return false;
```
Not strictly your fault, but I think you're exposing a bug in Socks5 here - Socks5 fails to CloseSocket() prior to returning false in one place (Proxy username or password too long), so I think it'd be better to just CloseSocket here instead of further down to make it harder to miss such things.

theuni replied:
Agreed.

laanwj replied:
+1 for the SocketCloser

TheBlueMatt commented on src/netbase.cpp:
```diff
@@ -445,6 +439,19 @@ bool static ConnectSocketDirectly(const CService &addrConnect, SOCKET& hSocketRe
+    struct sockaddr_storage sockaddr;
+    socklen_t len = sizeof(sockaddr);
+    if (!addrConnect.GetSockAddr((struct sockaddr*)&sockaddr, &len)) {
+        LogPrintf("Cannot connect to %s: unsupported network\n", addrConnect.ToString());
```
Need a CloseSocket here (but, really, lets just move all the CloseSockets up to OpenNetworkConnection post-CreateSocket instead of them being sprinkled all over everywhere?).

theuni replied:
Ugh. I'll see how much work it would be to RAII it as part of this pull.

TheBlueMatt reviewed (commented):
Review turned up a few bugs that should be fixed, though most of them not your fault :/.

theuni commented on src/net.cpp:
```diff
@@ -340,71 +340,161 @@ bool CConnman::CheckIncomingNonce(uint64_t nonce)
+    }
 
-        addrman.Attempt(addrConnect, fCountFailure);
+    if (!strDest.empty() && addrDest.IsValid()) {
```
I avoided coalescing on purpose. The current code does a bunch of steps at once, depending on what the input is. That's what I'm trying to undo here.

The previous block handles resolving, and will be moved out to a new function as a next step. By keeping the connection flow separated like this, it's trivial to factor out that way.

TheBlueMatt replied:
Ahh, OK, didnt realize that the next step would be to pull it out.

ryanofsky replied:
> By keeping the connection flow separated like this, it's trivial to factor out that way.

Probably there is something going over my head, but I don't understand the reasoning.

Currently this does:

```
if (condition) {
  // bunch of code
} else {
  // other code
}
if (equivalent condition) {
  // code that you want to factor out
}
if (equivalent condition) {
  // more code that you want to factor out
}
```

I don't see why it would be any harder to refactor later with:

```
if (condition) {
  // bunch of code
  // code that you want to factor out
  // more code that you want to factor out
} else {
  // other code
}
```

theuni replied:
While it may make sense to lump a few actions together if they meet the same conditions now because a connection is just one long synchronous operation, my intention is to clarify which conditions really need to be met at which stage, so that we can break that operation up into smaller asynchronous ones.

Some of these tests (like whether or not we resolved successfully) will just move in an obvious way to callbacks, so I guess it makes sense to combine them here. Others though, will need to be re-checked on separate threads.

I'll clarify this in the code by setting some variables rather than repeating conditions, so that it's more clear what the intent is.

I'll rebase while I'm at it.

theuni commented on src/net.cpp:
```diff
@@ -340,19 +340,29 @@ bool CConnman::CheckIncomingNonce(uint64_t nonce)
+    if (!fNetworkActive) {
+        return false;
+    }
+    if (IsLocal(addrConnect)) {
```
yes, thanks. Will fix.

theuni committed:
net: Add a quick RAII socket closer

theuni committed:
squashme: review fixups

- Only check IsLocal if pszDest is unset
- Fix function name string in CreateSocket

TheBlueMatt commented on src/net.cpp:
```diff
@@ -340,71 +340,160 @@ bool CConnman::CheckIncomingNonce(uint64_t nonce)
-        addrman.Attempt(addrConnect, fCountFailure);
+    // Create socket and make sure that it is closed in the event of an error
+    SOCKET hSocket = INVALID_SOCKET;
+    CSocketCloser sockCloser(hSocket);
```
It seems strange to create the CSocketCloser and then give it a reference to the socket - why not just have the CSocketCloser hold the SOCKET variable (publicly) and then set it to INVALID_SOCKET upon close/release (and return it from release) to make it feel more like a unique_ptr?

ryanofsky replied:
> It seems strange to create the CSocketCloser and then give it a reference to the socket

It is strange, just like CloseSocket taking a reference is strange. But I'd think as long as one of these is taking a reference, the other probably should as well for consistency.

theuni replied:
I gave it a reference because several calls further down the chain set the value to -1. For example, if ConnectSocketDirectly() fails. If storing a value, CSocketCloser would attempt to close it again at destruction if we weren't careful about resetting it, which would defeat the point of using RAII.

The alternative would be to audit for CloseSocket()s (or make a full-blown RAII CSocket with Close() as a member), which would make for a much larger diff.

ryanofsky reviewed (commented):
utACK 1d94e2ebe470cb8842223520a0c57513ca77e940. 

Would be nice to squash review fixes into place.

I don't think this PR is a pure improvement because `OpenNetworkConnection` is now a big monolith, but there are nice changes here and presumably there will be more cleanup in the future.

theuni commented:
Closing this for now. I'm working on getting this fully refactored with libevent instead.

Breaking this out and serializing the logic was very helpful as an intermediary step, but I now agree that it doesn't make much sense to review/merge those changes alone.
